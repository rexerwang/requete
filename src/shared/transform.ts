import type { RequestBody } from 'requete'

export function transformRequestBody(data?: RequestBody) {
  const toString = Object.prototype.toString.call(data)

  let requestBody: any = data
  let contentType = ''

  switch (toString) {
    case '[object Undefined]':
    case '[object Null]':
      break
    case '[object String]':
    case '[object URLSearchParams]':
      contentType = 'application/x-www-form-urlencoded;charset=utf-8'
      break
    case '[object FormData]':
    case '[object Blob]':
    case '[object ReadableStream]':
    case '[object ArrayBuffer]':
    case '[object DataView]':
    case '[object BigInt64Array]':
    case '[object BigUint64Array]':
    case '[object Float32Array]':
    case '[object Float64Array]':
    case '[object Int8Array]':
    case '[object Int16Array]':
    case '[object Int32Array]':
    case '[object Uint8Array]':
    case '[object Uint8ClampedArray]':
    case '[object Uint16Array]':
    case '[object Uint32Array]':
      contentType = 'multipart/form-data;charset=utf-8'
      break
    case '[object Object]':
    case '[object Array]':
      requestBody = JSON.stringify(data)
      contentType = 'application/json;charset=utf-8'
      break
    default:
      throw new TypeError('Invalid request body type: ' + toString)
  }

  return { requestBody, contentType }
}
