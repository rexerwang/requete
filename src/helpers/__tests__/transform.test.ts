import { transformRequestBody } from '../transform'

describe('transform specs', () => {
  it.each([
    undefined,
    null,
    '1',
    new FormData(),
    new Blob(['1']),
    new ArrayBuffer(1),
    { a: 1 },
    [{ a: 1 }],
  ])('should transform requestBody correctly when given `%o`', (data) => {
    expect(transformRequestBody(data)).toMatchSnapshot()
  })

  it('should thrown TypeError when given a number', () => {
    expect(() => transformRequestBody(1 as any)).toThrow(TypeError)
    expect(() => transformRequestBody(1 as any)).toThrow(
      /Invalid request body type/
    )
  })
})
