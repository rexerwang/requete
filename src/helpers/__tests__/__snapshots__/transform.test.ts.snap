// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform specs > should transform requestBody correctly when given \`'1'\` 1`] = `
{
  "contentType": "application/x-www-form-urlencoded;charset=utf-8",
  "requestBody": "1",
}
`;

exports[`transform specs > should transform requestBody correctly when given \`{ a: 1 }\` 1`] = `
{
  "contentType": "application/json;charset=utf-8",
  "requestBody": "{\\"a\\":1}",
}
`;

exports[`transform specs > should transform requestBody correctly when given \`{ a: 1 }\` 2`] = `
{
  "contentType": "application/json;charset=utf-8",
  "requestBody": "[{\\"a\\":1}]",
}
`;

exports[`transform specs > should transform requestBody correctly when given \`ArrayBuffer { [Uint8Contents]: <00>, byteLength: 1 }\` 1`] = `
{
  "contentType": "multipart/form-data;charset=utf-8",
  "requestBody": ArrayBuffer [],
}
`;

exports[`transform specs > should transform requestBody correctly when given \`Blob { size: 1, type: '' }\` 1`] = `
{
  "contentType": "multipart/form-data;charset=utf-8",
  "requestBody": Blob {
    Symbol(kHandle): Blob {},
    Symbol(kLength): 1,
    Symbol(kType): "",
  },
}
`;

exports[`transform specs > should transform requestBody correctly when given \`FormData {
  [Symbol(state)]: [ [length]: 0 ],
  [Symbol(Symbol.toStringTag)]: 'FormData'
}\` 1`] = `
{
  "contentType": "multipart/form-data;charset=utf-8",
  "requestBody": FormData {
    Symbol(state): [],
  },
}
`;

exports[`transform specs > should transform requestBody correctly when given \`null\` 1`] = `
{
  "contentType": "",
  "requestBody": null,
}
`;

exports[`transform specs > should transform requestBody correctly when given \`undefined\` 1`] = `
{
  "contentType": "",
  "requestBody": undefined,
}
`;
